runtime/connectors/blob/blobdownloader.go

```diff
@@ -1,94 +1,385 @@
 package blob
 
 import (
+	"container/list"
 	"context"
 	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
 
 	"cloud.google.com/go/storage"
 	"github.com/bmatcuk/doublestar/v4"
+	"github.com/c2h5oh/datasize"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/connectors"
 	"github.com/rilldata/rill/runtime/pkg/fileutil"
 	"gocloud.dev/blob"
+	"golang.org/x/sync/errgroup"
 )
 
+// increasing this limit can increase speed ingestion
+// but may increase bottleneck at duckdb or network/db IO
+// set without any benchamarks
+const concurrentBlobDownloadLimit = 8
+
+// map of supoprted extensions for partial downloads vs readers
+var partialDownloadReaders = map[string]string{".parquet": "parquet", ".csv": "csv", ".tsv": "csv", ".txt": "csv", ".parquet.gz": "parquet"}
+
+type Strategy string
+
+const (
+	TAIL Strategy = "tail"
+	HEAD Strategy = "head"
+	NONE Strategy = "none"
+)
+
+type ExtractConfig struct {
+	Strategy Strategy
+	Size     int64
+}
+
+type ExtractPolicy struct {
+	Row  ExtractConfig
+	File ExtractConfig
+}
```

Comment:
1. I did think about it. The decision to parse this is essentially the answer to, where do we want to keep our defaults. 

a) If we want to keep our defaults to the place where we intend to use these configs, current implementation holds good.
b) If we want to keep our defaults at one place at input parsing itself, we should change to what you are suggesting. 

Some cons of approach b) : 
1. Keeping defaults at input parsing itself and storing them in internal representation will mean that these will be exposed to the users as well when they check their source yamls(which may or may not necessarily be a bad thing). 
These defaults will also get locked into the artifacts permanently (which again may or may not necessarily be a bad thing). 
2. It may also not be possible to always set default values at input parsing itself 
    example : different defaults for different source connectors(gcs vs s3)

Please let me know your thoughts. I am open to choosing both approaches.

Reply:
I think I misunderstood the comment. Your comment just talks about parsing and not about keeping any defaults. I will change that.
Only issue that I see with doing this in input parsing is that we will never be able to guarantee that we output source yaml to same as what user chooses. For example there are chances that user choose 1024 MB as size but while converting it from proto to yaml, it outputs 1GB. I don't see any issue with doing this though.

Reply:
Got it. To clarify, we actually want to avoid ever having to output YAML. The user creates the YAML (maybe through the frontend), and the runtime just consumes it. There are some functions to output YAML for an object, but it's just for testing. So no need to worry about the reverse direction (at the moment), just `code files -> internal repr`.